framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,726 @@
+     *     parameter, then the type of the respective value passed as argument
+     *     in the method call methodInvNode will be added to the parameter in
+     *     the Scene.</li>
+     *     <li>If the Scene previously contained an annotated type for that
+     *     parameter, then its new type will be the LUB between the previous
+     *     type and the type of the respective value passed as argument in the
+     *     method call.</li>
+     *   </ul>
+     * <p>
+     * @param methodInvNode the node representing a method invocation.
+     * @param receiverTree the Tree of the class that contains the method being
+     * invoked.
+     * @param methodElt the element of the method being invoked.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method parameters' types.
+     */
+    public static void updateInferredMethodParametersTypes(
+            MethodInvocationNode methodInvNode, Tree receiverTree,
+            ExecutableElement methodElt, AnnotatedTypeFactory atf) {
+        if (receiverTree == null) {
+            // TODO: Method called from static context.
+            // I struggled to obtain the ClassTree of a method called
+            // from a static context and currently I'm ignoring it.
+            return;
+        }
+        ClassSymbol classSymbol = getEnclosingClassSymbol(receiverTree);
+        if (classSymbol == null) {
+            // TODO: Handle anonymous classes.
+            // Also struggled to obtain the ClassTree from an anonymous class.
+            // Ignoring it for now.
+            return;
+        }
+        String className = classSymbol.flatname.toString();
+
+        String jaifPath = getJaifPath(className);
+        AClass clazz = getAClass(className, jaifPath);
+
+        String methodName = JVMNames.getJVMMethodName(methodElt);
+        AMethod method = clazz.methods.vivify(methodName);
+
+        for (int i = 0; i < methodInvNode.getArguments().size(); i++) {
+            VariableElement ve = methodElt.getParameters().get(i);
+            AnnotatedTypeMirror paramATM = atf.getAnnotatedType(ve);
+
+            Node arg = methodInvNode.getArgument(i);
+            Tree treeNode = arg.getTree();
+            if (treeNode == null) {
+                // TODO: Handle variable-length list as parameter.
+                // An ArrayCreationNode with a null tree is created when the
+                // parameter is a variable-length list. We are ignoring it for now.
+                continue;
+            }
+            AnnotatedTypeMirror argATM = atf.getAnnotatedType(treeNode);
+            AField param = method.parameters.vivify(i);
+            updateAnnotationSetInScene(
+                    param.type, atf, jaifPath, argATM, paramATM,
+                    DefaultLocation.PARAMETERS);
+        }
+    }
+
+    /**
+     * Updates the type of the field lhs in the Scene of the class with
+     * symbol classSymbol.
```

Comment:
The Javadoc is strange because it refers to a local variable classSymbol. Someone reading the Javadoc by itself would not see the declaration of classSymbol. Would it be better to describe it in terms of classTree?